### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by SargtLa**

### CODE REVIEW

File: api/company.go [hunk 1]
```diff
@@ -170,11 +171,26 @@ func EmptyValue(value interface{}) bool {
...
-	case int64:
+	case int32, int64, float32, float64:
 		return val == 0
+	case time.Time:
+		return val.IsZero()
+	case *time.Time:
+		if val == nil {
+			return true
+		} else {
+			return (*val).IsZero()
```
Comment (ruslanBik4) [inline api/company.go]: val.IsZero()

File: db/consts.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
+package db
+
+// Table names
```
Comment (ruslanBik4) [inline db/consts.go:5]: Consts must be group & their names must have upper_case (like TABLE_TAGS)

File: db/init_db.go [hunk 3]
```diff
@@ -33,6 +33,23 @@ func GetDB(ctxApis apis.CtxApis) *dbEngine.DB {
...
 	}
 
+	tagIds = &TagIdMap{}
+	tagsTable, err := NewTags(db)
+	if err != nil {
+		logs.ErrorLog(err, "cannot get %s table", TableTags)
+		return db
+	}
+
+	err = tagsTable.SelectSelfScanEach(ctx,
```
Comment (ruslanBik4) [inline db/init_db.go]: May be there SelectToMap works well?

File: db/tags.go [hunk 4]
```diff
@@ -170,3 +174,113 @@ func (t *Tags) Update(ctx context.Context, Options ...dbEngine.BuildSqlOptions)
 
 	return t.Table.Update(ctx, Options...)
 }
+
+func GetTagIdFirstContact() int32 {
+	if tag, ok := (*tagIds)[TagFirstContact]; ok {
```
Comment (ruslanBik4) [inline db/tags.go]: tagIds[TagFirstContact]

File: httpgo.go [hunk 5]
```diff
@@ -55,6 +55,8 @@ var (
 
 var httpServer *httpgo.HttpGo
 
+var TagIds *db.TagIdMap
```
Comment (ruslanBik4) [inline httpgo.go]: rm it from main module

**Review by ruslanBik4 — CHANGES_REQUESTED** [review]

File: db/tags.go (hunk 4 above)
Comment (SargtLa) [reply db/tags.go]: dont work that way

File: db/init_db.go (hunk 3 above)
Comment (SargtLa) [reply db/init_db.go]: why not ScanEach?

File: db/consts.go (hunk 2 above)
Comment (SargtLa) [reply db/consts.go:5]: consts

File: db/init_db.go (hunk 3 above)
Comment (SargtLa) [reply db/init_db.go]: If you want to use SelectToMap, you need to call it from connection, set sql request... why do we need all that if we've already got a table anyway?

File: db/consts.go (hunk 2 above)
Comment (SargtLa) [reply db/consts.go:5]: so what's your decision?

File: api/helpers.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
+package api
```
Comment (ruslanBik4) [inline api/helpers.go]: wrong name of module - "helpers'

File: db/candidates.go [hunk 7]
```diff
@@ -216,9 +216,9 @@ func (r *CandidatesFields) ColValue(name string) interface{} {
 }
 
 func NewCandidates(db *dbEngine.DB) (*Candidates, error) {
-	table, ok := db.Tables["candidates"]
+	table, ok := db.Tables[TABLE_Candidates]
```
Comment (ruslanBik4) [inline db/candidates.go]: name of constant must be in Upper_Case (TABLE_CANDIDATES)

File: db/form_blocks.go [hunk 8]
```diff
@@ -82,9 +82,9 @@ func (r *Form_blocksFields) ColValue(name string) interface{} {
 }
 
 func NewForm_blocks(db *dbEngine.DB) (*Form_blocks, error) {
-	table, ok := db.Tables["form_blocks"]
+	table, ok := db.Tables[TABLE_FormBlocks]
```
Comment (ruslanBik4) [inline db/form_blocks.go]: TABLE_FORM_BLOKCS

File: db/init_db.go [hunk 9]
```diff
@@ -243,3 +249,125 @@ func getOidCustomTypes(ctx context.Context, conn *pgx.Conn) error {
...
+	statusesIds = StatusIdMap{}
+	statusesTable, err := NewStatuses(db)
+	if err != nil {
+		logs.ErrorLog(err, "cannot get %s table", TABLE_Statuses)
+		return err
+	}
+
+	err = statusesTable.SelectSelfScanEach(ctx,
+		func(record *StatusesFields) error {
+			statusesIds[record.Status.String] = *record
```
Comment (ruslanBik4) [inline db/init_db.go]: change field Status into type string

File: db/init_db.go [hunk 10]
```diff
@@ -243,3 +249,125 @@ func getOidCustomTypes(ctx context.Context, conn *pgx.Conn) error {
...
+	statusesForVacIds = StatusForVacIdMap{}
+	statusesForVacsTable, err := NewStatusForVacs(db)
+	if err != nil {
+		logs.ErrorLog(err, "cannot get %s table", TABLE_StatusForVacs)
+		return err
+	}
+
+	err = statusesForVacsTable.SelectSelfScanEach(ctx,
+		func(record *StatusForVacsFields) error {
+			statusesForVacIds[record.Status.String] = *record
```
Comment (ruslanBik4) [inline db/init_db.go]: type string

File: db/init_db.go [hunk 11]
```diff
@@ -243,3 +249,125 @@ func getOidCustomTypes(ctx context.Context, conn *pgx.Conn) error {
...
+	seniorityIds = SeniorityIdMap{}
+	seniorityTable, err := NewSeniorities(db)
+	if err != nil {
+		logs.ErrorLog(err, "cannot get %s table", TABLE_Seniorities)
+		return err
+	}
+
+	err = seniorityTable.SelectSelfScanEach(ctx,
+		func(record *SenioritiesFields) error {
+			seniorityIds[record.Nazva.String] = *record
```
Comment (ruslanBik4) [inline db/init_db.go]: rename nazva to name & chg type to string

File: db/init_db.go [hunk 12]
```diff
@@ -243,3 +249,125 @@ func getOidCustomTypes(ctx context.Context, conn *pgx.Conn) error {
...
+	platformIds = PlatformsIdMap{}
+	platformsTable, err := NewPlatforms(db)
+	if err != nil {
+		logs.ErrorLog(err, "cannot get %s table", TABLE_Platforms)
+		return err
+	}
+
+	err = platformsTable.SelectSelfScanEach(ctx,
+		func(record *PlatformsFields) error {
+			platformIds[record.Nazva.String] = *record
```
Comment (ruslanBik4) [inline db/init_db.go]: rename nazva to name & chg type to string

File: db/platforms.go [hunk 13]
```diff
@@ -130,3 +133,29 @@ func (t *Platforms) Update(ctx context.Context, Options ...dbEngine.BuildSqlOpti
...
+
+	return nil
+}
+
+func GetPlatformsAsSelectedUnits() (res SelectedUnits) {
+	if len(platformIds) == 0 {
+		return nil
+	}
+
+	for _, platform := range platformIds {
```
Comment (ruslanBik4) [inline db/platforms.go]: res SelectedUnits must be construct one time

File: db/seniorities.go [hunk 14]
```diff
@@ -130,3 +135,108 @@ func (t *Seniorities) Update(ctx context.Context, Options ...dbEngine.BuildSqlOp
...
+	}
+	return nil
+}
+
+func GetSenioritiesAsSelectedUnits() (res SelectedUnits) {
+	if len(seniorityIds) == 0 {
+		return nil
+	}
+
+	for _, sen := range seniorityIds {
```
Comment (ruslanBik4) [inline db/seniorities.go]: one time construct

File: db/status_for_vacs.go [hunk 15]
```diff
@@ -144,3 +149,119 @@ func (t *Status_for_vacs) Update(ctx context.Context, Options ...dbEngine.BuildS
 
 	return t.Table.Update(ctx, Options...)
 }
+
+func GetStatusForVacIdInterview() int32 {
+	if statForVac, ok := statusesForVacIds[STATUS_FOR_VAC_Interview]; ok {
+		return statForVac.Id
+	} else {
+		logs.ErrorLog(errors.Errorf("StatusForVac \"%s\" not found in database", STATUS_FOR_VAC_Interview))
```
Comment (ruslanBik4) [inline db/status_for_vacs.go]: errors case must be first

**Review by ruslanBik4 — CHANGES_REQUESTED** [review]

File: api/helpers.go (hunk 6 above)
Comment (SargtLa) [reply api/helpers.go]: why?

File: db/platforms.go (hunk 13 above)
Comment (SargtLa) [reply db/platforms.go]: didn't get it

**Pull request merged**